### DISCUSSION TITLE
Is posible to integrate keys from pam_google_authenticator.so

### DISCUSSION THREAD

**Discussion by crebollobr** [authelia/authelia#6763] [Q&A] [answered]
Hello, 

We have 1k of users with acount in Linux using pam_google_authenticator.so 2fa that save the key in user home.
Is there any way to impot those files? 
The users account are in LDAP.

We have a test site running authelia authenticating in ldap but with new key

Thanks

**Comment by glime-ai** [comment]
Please note this response was created using generative AI leveraging previously available information from the repository. This information may be helpful however it is our strong recommendation that you carefully consider the information before blindly applying it.

---

To import existing 2FA keys from `pam_google_authenticator` for use with Authelia, you would need to undertake a custom migration process, as Authelia does not natively support importing 2FA keys directly from external systems or files. Authelia's primary method for 2FA involves users registering their devices through its own interface, as detailed in the provided documentation. Specifically, for TOTP (Time-based One-Time Password), users must register their device by scanning a QR code generated by Authelia, which does not directly accommodate importing pre-existing TOTP keys from `pam_google_authenticator`.

Given your scenario, where user accounts are managed via LDAP and you have a significant number of users with existing 2FA configurations, a possible approach would involve:

1. **Extracting TOTP Secrets**: Develop a script or process to extract the TOTP secrets from the `pam_google_authenticator` files located in each user's home directory.

2. **Formatting Data for Authelia**: Convert the extracted TOTP secrets into a format that can be recognized by Authelia. This might involve creating a structured file or database entries depending on how you plan to manage user 2FA tokens within Authelia.

3. **Automating Token Registration**: Since Authelia does not provide a direct method to import these tokens, you might need to simulate the token registration process or directly manipulate Authelia's storage backend to insert the TOTP secrets. This approach requires in-depth knowledge of Authelia's storage schema and careful consideration of security implications.

4. **Testing and Validation**: Before rolling out any changes to your production environment, thoroughly test the migration process to ensure that it works as expected and does not compromise the security of your 2FA setup.

5. **User Communication**: Inform your users about the migration and any actions they might need to take, such as verifying their 2FA setup or re-registering their devices if necessary.

It's important to note that this approach involves several technical challenges and potential security risks. Direct manipulation of Authelia's data storage or attempting to automate the 2FA registration process should be done with caution. Additionally, this solution might require custom development work and is not supported out of the box by Authelia.

For assistance with Authelia's configuration options, including LDAP integration and 2FA methods, you can refer to the official documentation and possibly reach out to the Authelia community or support channels for advice on handling such a migration.

This comment was generated by [Glime](https://glimelab.ai/?utm_source=6fsHkcyufjQLaB&utm_medium=github).

**Comment by james-d-elliott** [comment]
Are you able to create a test user, setup TOTP, and show the contents of the file for that user? I would imagine it's possible.

**Reply by crebollobr** [reply]
Test user

The first string is the key i cofigure in mobile app 

G2SXWSQAPOEQ2ZDZQEBYDZNSZU
" RATE_LIMIT 3 30
" WINDOW_SIZE 17
" DISALLOW_REUSE
" TOTP_AUTH
72561028
34174211
30590044
55488970
45139672

**Reply by crebollobr** [reply]
In Authelia
root@note360:~# ./authelia-linux-amd64 storage user totp export --format csv --config config.yml
issuer,username,algorithm,digits,period,secret
teste.com.br,rebollo,SHA1,6,30,JQGT4KTEM3MPIEL5AVKKQJ5MDKJERPX3XPQTUOTFOU3C3XLL2WBQ

**Comment by crebollobr** [comment]
There is authelia storage user totp export
but not authelia storage user totp import

**Reply by james-d-elliott** [reply, marked as answer]
There's a very soon to be released version which has a YAML export which can be imported again.

**Comment by nightah** [comment]
I also believe that a user created a script in the reverse which would pull TOTP secrets out of Authelia and insert them for usage with `pam_google_authenticator`, it may have been @davama if my memory serves correctly.
